packages/aws-lambda/src/ssm.js

```diff
@@ -0,0 +1,143 @@
+
+  try {
+    /**
+     * From https://docs.aws.amazon.com/lambda/latest/dg/lambda-nodejs.html:
+     *
+     * Your code runs in an environment that includes the AWS SDK for JavaScript,
+     * with credentials from an AWS Identity and Access Management (IAM) role that you manage.
+     */
+    // eslint-disable-next-line
+    AWS = require('aws-sdk');
+
+    // https://docs.aws.amazon.com/lambda/latest/dg/configuration-envvars.html
+    AWS.config.update({ region: process.env.AWS_REGION });
+
+    const ssm = new AWS.SSM();
+    const params = {
+      Name: envValue,
+      // See https://docs.aws.amazon.com/cli/latest/reference/ssm/get-parameter.html#options
+      // We always expect a string here. Never a decrypt
+      WithDecryption: false
+    };
+
+    logger.debug(`INSTANA_SSM_PARAM_NAME is ${envValue}.`);
+
+    ssm.getParameter(params, (err, data) => {
+      if (err) {
+        errorFromAWS = `Error from AWS-SDK SSM Parameter Store: ${err.message}`;
+      } else {
+        try {
+          fetchedValue = data.Parameter.Value;
+          errorFromAWS = null;
+          logger.debug(`INSTANA AGENT KEY: ${fetchedValue}`);
+        } catch (readError) {
+          errorFromAWS = `Could not read returned response from AWS-SDK SSM Parameter Store: ${readError.message}`;
+        }
+      }
+    });
+  } catch (err) {
+    logger.warn('AWS SDK not available.');
+    errorFromAWS =
+      'Could not fetch instana key from SSM parameter store using ' +
+      `"${process.env.INSTANA_SSM_PARAM_NAME}", because the AWS SDK is not available. ` +
+      `Reason: ${err.message}`;
+  }
+};
+
+exports.waitAndGetInstanaKey = callback => {
+  // CASE: We already fetched the SSM value, skip & save time
+  if (fetchedValue) {
+    return callback(null, fetchedValue);
+  }
+  // CASE: Customer has set INSTANA_SSM_PARAM_NAME, but we were not able to fetch the value from AWS
+  if (errorFromAWS) {
+    return callback(errorFromAWS);
+  }
+
+  const endInMs = Date.now();
+  // CASE: the time between ssm lib initialisation and waitAndGetInstanaKey call
+  //       (which is the end of the customers lambda handler) is already too big to wait for the AWS response
+  if (endInMs - initTimeoutInMs > (process.env.INSTANA_LAMBDA_SSM_AWS_TIMEOUT_IN_MS || 1000)) {
+    return callback('Stopped waiting for AWS SSM response.');
+  }
+
+  /**
+   * Inside AWS the call to `getParameter` mostly takes 30-50ms
+   * Because we initialise the fetch already before the customer's handler is called,
+   * the chance is very high that the interval is not even used.
+   *
+   * In our tests it takes usually ~>150ms (remote call)
+   */
+  const stopIntervalAfterMs = process.env.INSTANA_LAMBDA_SSM_TIMEOUT_IN_MS || 150;
+  const start = Date.now();
+
+  const waiting = setInterval(() => {
+    const end = Date.now();
+
+    if (fetchedValue) {
+      clearInterval(waiting);
+
+      callback(null, fetchedValue);
+    } else if (end - start > stopIntervalAfterMs) {
```

Comment:
Maybe `stopIntervalAfterMs` should have a max value, as it can be set by customers through `process.env.INSTANA_LAMBDA_SSM_TIMEOUT_IN_MS`.
If the value is too high, it could cause the setInterval ticks to run for a long period of time.
I don't have enough experience with AWS Lambda, but couldn't this hold the process for too long or cause some kind of unwanted behavior if the interval ticks continue for too long? I believe that "too long" here means 1 or 2 seconds, or something like this, no?

Reply:
Or maybe we can have a third condition that guarantees that the setInterval tick loop ends at some point.

Reply:
> as it can be set by customers through process.env.INSTANA_LAMBDA_SSM_TIMEOUT_IN_MS.

Same as above :) Will switch to function options. Customer should not set these values